cpp/include/raft/matrix/detail/gather.cuh

```diff
@@ -343,6 +349,93 @@ void gather_if(const InputIteratorT in,
   typedef typename std::iterator_traits<MapIteratorT>::value_type MapValueT;
   gatherImpl(in, D, N, map, stencil, map_length, out, pred_op, transform_op, stream);
 }
+
+template <typename MatrixT, typename MapT, typename MapTransformOp, typename IndexT>
+void gatherInplaceImpl(raft::resources const& handle,
+                       raft::device_matrix_view<MatrixT, IndexT, raft::layout_c_contiguous> inout,
+                       raft::device_vector_view<const MapT, IndexT, raft::layout_c_contiguous> map,
+                       MapTransformOp transform_op,
+                       IndexT batch_size)
+{
+  // return type of MapTransformOp, must be convertible to IndexT
+  typedef typename std::result_of<decltype(transform_op)(MapT)>::type MapTransformOpReturnT;
+  RAFT_EXPECTS((std::is_convertible<MapTransformOpReturnT, IndexT>::value),
+               "MapTransformOp's result type must be convertible to signed integer");
+
+  IndexT m          = inout.extent(0);
+  IndexT n          = inout.extent(1);
+  IndexT map_length = map.extent(0);
+
+  // skip in case of 0 length input
+  if (map_length <= 0 || m <= 0 || n <= 0 || batch_size < 0) return;
+
+  RAFT_EXPECTS(map_length <= m, "Length of map should be <= number of rows for inplace gather");
+
+  // re-assign batch_size for default case
+  if (batch_size == 0) batch_size = n;
+
+  RAFT_EXPECTS(batch_size <= n, "batch size should be <= number of columns");
+
+  auto exec_policy = resource::get_thrust_policy(handle);
+  IndexT n_batches = raft::ceildiv(n, batch_size);
+  for (IndexT bid = 0; bid < n_batches; bid++) {
+    IndexT batch_offset   = bid * batch_size;
+    IndexT cols_per_batch = min(batch_size, n - batch_offset);
+    auto scratch_space =
+      raft::make_device_vector<MatrixT, IndexT>(handle, map_length * cols_per_batch);
+
+    auto gather_op = [inout = inout.data_handle(),
+                      map   = map.data_handle(),
```

Comment:
I really like this pattern where the closure is explicitly shadowing parent variables with views. Nice!